packages/astro/src/assets/build/generate.ts

```diff
@@ -53,113 +72,176 @@ export async function generateImage(
+	env: AssetEnv,
+	queue: PQueue
+) {
+	const originalImageData = await loadImage(originalFilePath, env);
+
+	for (const [_, transform] of transforms) {
+		queue.add(async () =>
+			generateImage(originalImageData, transform.finalPath, transform.transform)
+		);
+	}
 
-	// For remote images, instead of saving the image directly, we save a JSON file with the image data and expiration date from the server
-	const cacheFile = basename(filepath) + (isLocalImage ? '' : '.json');
-	const cachedFileURL = new URL(cacheFile, assetsCacheDir);
+	async function generateImage(
+		originalImage: ImageData,
+		filepath: string,
+		options: ImageTransform
+	) {
+		let timeStart = performance.now();
+		const generationData = await generateImageInternal(originalImage, filepath, options);
 
-	await fs.promises.mkdir(finalFolderURL, { recursive: true });
+		if (!generationData) {
+			return;
+		}
 
-	// Check if we have a cached entry first
-	try {
-		if (isLocalImage) {
-			await fs.promises.copyFile(cachedFileURL, finalFileURL);
+		const timeEnd = performance.now();
+		const timeChange = getTimeStat(timeStart, timeEnd);
+		const timeIncrease = `(+${timeChange})`;
+		const statsText = generationData.cached
+			? `(reused cache entry)`
+			: `(before: ${generationData.weight.before}kB, after: ${generationData.weight.after}kB)`;
+		const count = `(${env.count.current}/${env.count.total})`;
+		env.logger.info(
+			null,
+			`  ${green('▶')} ${filepath} ${dim(statsText)} ${dim(timeIncrease)} ${dim(count)}`
+		);
+		env.count.current++;
+	}
 
-			return {
-				cached: true,
-			};
-		} else {
-			const JSONData = JSON.parse(readFileSync(cachedFileURL, 'utf-8')) as RemoteCacheEntry;
+	async function generateImageInternal(
+		originalImage: ImageData,
+		filepath: string,
+		options: ImageTransform
+	): Promise<GenerationData | undefined> {
+		const isLocalImage = isESMImportedImage(options.src);
+		const finalFileURL = new URL('.' + filepath, env.clientRoot);
 
-			// If the cache entry is not expired, use it
-			if (JSONData.expires > Date.now()) {
-				await fs.promises.writeFile(finalFileURL, Buffer.from(JSONData.data, 'base64'));
+		// For remote images, instead of saving the image directly, we save a JSON file with the image data and expiration date from the server
+		const cacheFile = basename(filepath) + (isLocalImage ? '' : '.json');
+		const cachedFileURL = new URL(cacheFile, env.assetsCacheDir);
+
+		// Check if we have a cached entry first
+		try {
+			if (isLocalImage) {
+				await fs.promises.copyFile(cachedFileURL, finalFileURL);
 
 				return {
 					cached: true,
 				};
-			}
-		}
-	} catch (e: any) {
-		if (e.code !== 'ENOENT') {
-			throw new Error(`An error was encountered while reading the cache file. Error: ${e}`);
-		}
-		// If the cache file doesn't exist, just move on, and we'll generate it
-	}
+			} else {
+				const JSONData = JSON.parse(readFileSync(cachedFileURL, 'utf-8')) as RemoteCacheEntry;
```

Comment:
We should handle the error thrown by `JSON.parse`. You never know if someone tampered with the file.

Reply:
We do, this code is in a `try catch`.